iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/GenericUDFIcebergBucket.java

```diff
@@ -0,0 +1,198 @@
+        DecimalTypeInfo decimalTypeInfo = (DecimalTypeInfo) TypeInfoUtils.getTypeInfoFromObjectInspector(argumentOI);
+        Type.PrimitiveType decimalIcebergType = Types.DecimalType.of(decimalTypeInfo.getPrecision(),
+            decimalTypeInfo.getScale());
+
+        converter = new PrimitiveObjectInspectorConverter.HiveDecimalConverter(argumentOI,
+            PrimitiveObjectInspectorFactory.writableHiveDecimalObjectInspector);
+        evaluator = arg -> {
+          HiveDecimalWritable val = (HiveDecimalWritable) converter.convert(arg.get());
+          applyBucketTransform(val.getHiveDecimal().bigDecimalValue(), decimalIcebergType);
+        };
+        break;
+
+      case FLOAT:
+        converter = new PrimitiveObjectInspectorConverter.FloatConverter(argumentOI,
+            PrimitiveObjectInspectorFactory.writableFloatObjectInspector);
+        evaluator = arg -> {
+          FloatWritable val = (FloatWritable) converter.convert(arg.get());
+          applyBucketTransform(val.get(), Types.FloatType.get());
+        };
+        break;
+
+      case DOUBLE:
+        converter = new PrimitiveObjectInspectorConverter.DoubleConverter(argumentOI,
+            PrimitiveObjectInspectorFactory.writableDoubleObjectInspector);
+        evaluator = arg -> {
+          DoubleWritable val = (DoubleWritable) converter.convert(arg.get());
+          applyBucketTransform(val.get(), Types.DoubleType.get());
+        };
+        break;
+
+      default:
+        throw new UDFArgumentException(
+            " ICEBERG_BUCKET() only takes STRING/CHAR/VARCHAR/BINARY/INT/LONG/DECIMAL/FLOAT/DOUBLE" +
+                " types as first argument, got " + inputType);
+    }
+
+    numBuckets = getNumBuckets(arguments[1]);
+
+    outputOI = PrimitiveObjectInspectorFactory.writableIntObjectInspector;
+    return outputOI;
+  }
+
+  private static int getNumBuckets(ObjectInspector arg) throws UDFArgumentException {
+    UDFArgumentException udfArgumentException = new UDFArgumentException("ICEBERG_BUCKET() second argument can only " +
+        "take an int type, but got " + arg.getTypeName());
+    if (arg.getCategory() != ObjectInspector.Category.PRIMITIVE) {
+      throw udfArgumentException;
+    }
+    PrimitiveObjectInspector.PrimitiveCategory inputType = ((PrimitiveObjectInspector) arg).getPrimitiveCategory();
+    if (inputType != PrimitiveObjectInspector.PrimitiveCategory.INT) {
+      throw udfArgumentException;
+    }
+    return ((WritableConstantIntObjectInspector) arg).getWritableConstantValue().get();
+  }
+
+  @Override
+  public Object evaluate(DeferredObject[] arguments) throws HiveException {
+
+    DeferredObject argument = arguments[0];
+    if (argument == null) {
+      return null;
+    } else {
+      evaluator.apply(argument);
+    }
+
+    return result;
+  }
+
+  private <T> void applyBucketTransform(T value, Type.PrimitiveType icebergType) {
+    Transform<T, Integer> transform = Transforms.bucket(icebergType, numBuckets);
```

Comment:
Would it make sense to create the transform in the `initiaiize` method instead of recreating it for every row?

Reply:
Yes it would, I just refactored this. I must say I really liked my generic method though :/